Multiroot bindings

catompkins opened the pull request:
Added bindings for multiroot solvers.

- this required an update to `regen.rs` which caused subsequent changes to `auto.rs`. I used rust 1.54 + GSL v2.7 release. 
- the current auto.rs file required updates to `matrix.rs` and `vector.rs` due to mismatched type errors. You noted that there was a bug there previously w/ inconsistent types. It doesn't appear to be there at the moment.

GuillaumeGomez commented on src/types/matrix.rs:
```diff
@@ -270,19 +270,19 @@ impl $rust_name {
-    pub fn scale(&mut self, x: f64) -> Value {
+    pub fn scale(&mut self, x: $rust_ty) -> Value {
         Value::from(unsafe { sys::[<$name _scale>](self.unwrap_unique(), x) })
     }
 
     /// This function adds the constant value x to the elements of the self matrix. The result
     /// self(i,j) <- self(i,j) + x is stored in self.
     #[doc(alias = $name _add_constant)]
-    pub fn add_constant(&mut self, x: f64) -> Value {
+    pub fn add_constant(&mut self, x: $rust_ty) -> Value {
         Value::from(unsafe { sys::[<$name _add_constant>](self.unwrap_unique(), x) })
     }
 
     #[doc(alias = $name _add_constant)]
-    pub fn add_diagonal(&mut self, x: f64) -> Value {
+    pub fn add_diagonal(&mut self, x: $rust_ty) -> Value {
```
These changes seem invalid: https://github.com/rwinlib/gsl/blob/master/include/gsl/gsl_matrix_int.h#L250-L252

GuillaumeGomez commented on src/types/multiroot.rs:
```diff
@@ -0,0 +1,286 @@
+    }
+
+    /// This function returns the current estimate of the root for the solver `s`, given by `s->x`.
+    pub fn root(& self) -> ::VectorF64 {
```
```suggestion
    pub fn root(&self) -> ::VectorF64 {
```

GuillaumeGomez commented on src/types/multiroot.rs:
```diff
@@ -0,0 +1,286 @@
+    }
+
+    /// This function returns the last step `dx` taken by the solver `s`, given by `s->dx`.
+    pub fn dx(& self) -> ::VectorF64 {
```
```suggestion
pub fn dx(& self) -> ::VectorF64 {
```

GuillaumeGomez commented on src/types/multiroot.rs:
```diff
@@ -0,0 +1,286 @@
+
+    /// This function returns the function value `f(x)` at the current estimate of the root for
+    /// the solver `s`, given by `s->f`.
+    pub fn f(& self) -> ::VectorF64 {
```
```suggestion
    pub fn f(&self) -> ::VectorF64 {
```

GuillaumeGomez commented on src/types/multiroot.rs:
```diff
@@ -0,0 +1,286 @@
+
+    /// checking a test function
+    /// must return a success criteria (or failure)
+    fn rosenbrock_f(x: & ::VectorF64, f: &mut ::VectorF64) -> ::Value{
```
I'm surprised `cargo fmt` didn't catch this one...

```suggestion
    fn rosenbrock_f(x: &::VectorF64, f: &mut ::VectorF64) -> ::Value{
```

GuillaumeGomez commented on src/types/multiroot.rs:
```diff
@@ -0,0 +1,286 @@
+        }
+        assert!(matches!(status, ::Value::Success))
+    }
+}
```
Please add a backline at the end of the file.

catompkins replied:
added

GuillaumeGomez commented on src/types/vector.rs:
```diff
@@ -267,8 +267,7 @@ impl $rust_name {
     #[doc(alias = $name _add_constant)]
-    pub fn add_constant(&mut self, x: f64) -> Value {
-        // Funny bug: here it expects a f64 and not a f32 for gsl_vector_float...
+    pub fn add_constant(&mut self, x: $rust_ty) -> Value {
```
:+1:

GuillaumeGomez replied:
Surprising that the compiler was happy with that...

GuillaumeGomez commented on src/multiroot.rs:
```diff
@@ -0,0 +1,15 @@
+#[doc(alias = "gsl_multiroot_test_residual")]
+pub fn test_residual(f: &::VectorF64, epsabs: f64) -> ::Value {
+    ::Value::from(unsafe { sys::gsl_multiroot_test_residual(f.unwrap_shared(), epsabs) })
+}
```
Please end the file with a backline.

GuillaumeGomez commented on src/multiroot.rs:
```diff
@@ -0,0 +1,15 @@
+// A rust binding for the GSL library
+//
+//! Multiroot test algorithms, See `rgsl::types::multiroot` for solvers.
+use ffi::FFI;
```
Please add an empty line before this one.

GuillaumeGomez commented on src/multiroot.rs:
```diff
@@ -0,0 +1,15 @@
+//
+// A rust binding for the GSL library
+//
+//! Multiroot test algorithms, See `rgsl::types::multiroot` for solvers.
```
Please add an empty line before this one.

catompkins replied:
added

GuillaumeGomez commented on src/types/matrix.rs:
```diff
@@ -270,19 +270,19 @@ impl $rust_name {
     #[doc(alias = $name _add_constant)]
-    pub fn add_diagonal(&mut self, x: f64) -> Value {
+    pub fn add_diagonal(&mut self, x: $rust_ty) -> Value {
         Value::from(unsafe { sys::[<$name _add_constant>](self.unwrap_unique(), x) })
```
Just realized that `add_constant`  was incorrectly used! Sending a PR to fix that...

catompkins replied:
merged your changes into `catompkins:multiroot_binding`

GuillaumeGomez commented:
So, after seeing your PR, I decided it was time to clean things up. So in the end, I modified the sys generator to take the 2.7 release directly, fixed some C header issues, added a checker for a few things. You can see all these changes in https://github.com/GuillaumeGomez/rust-GSL/pull/118.

Sorry to come a bit ouf of the blue like but could you rebase on master please? Normally the diff should be **much** smaller now without the sys changes. Thanks for your contribution in any case!

catompkins commented:
> So, after seeing your PR, I decided it was time to clean things up. So in the end, I modified the sys generator to take the 2.7 release directly, fixed some C header issues, added a checker for a few things. You can see all these changes in #118.
> 
> Sorry to come a bit ouf of the blue like but could you rebase on master please? Normally the diff should be **much** smaller now without the sys changes. Thanks for your contribution in any case!

Rebased on master + updated auto.rs from regen.

GuillaumeGomez commented on src/multiroot.rs:
```diff
@@ -0,0 +1,19 @@
+use ffi::FFI;
+
+#[doc(alias = "gsl_multiroot_test_delta")]
+pub fn test_interval(dx: &::VectorF64, x: &::VectorF64, epsabs: f64, epsrel: f64) -> ::Value {
```
Shouldn't it be named `test_delta`?

GuillaumeGomez commented on src/types/multiroot.rs:
```diff
@@ -0,0 +1,294 @@
+    ) -> ::Value {
+        unsafe extern "C" fn inner_f<A: Fn(&::VectorF64, &mut ::VectorF64) -> ::Value>(
+            x: *const sys::gsl_vector,
+            params: *mut ::std::os::raw::c_void,
```
Please use libc types.

GuillaumeGomez commented on src/types/multiroot.rs:
```diff
@@ -0,0 +1,294 @@
+    #[doc(alias = "gsl_multiroot_fsolver_set")]
+    pub fn set<F: Fn(&::VectorF64, &mut ::VectorF64) -> ::Value>(
+        &mut self,
+        f: F,
```
You also need to store this closure otherwise it might be dropped before being used.

GuillaumeGomez commented on src/types/multiroot.rs:
```diff
@@ -0,0 +1,339 @@
+            x: *const sys::gsl_vector,
+            params: *mut c_void,
+            f: *mut sys::gsl_vector,
+        ) -> std::os::raw::c_int {
```
Please use libc types.

GuillaumeGomez commented:
Just one remaining nit and it's good to go. This is great, thanks a lot! Please fix it then rebase on top master. Once this PR is merged, I'll make a new release.

GuillaumeGomez commented:
I think you messed up your git history (why is there still 3b4c6e4 for example). :laughing: Please clean it up then I'll make a last review.

catompkins commented:
> I think you messed up your git history (why is there still [3b4c6e4](https://github.com/GuillaumeGomez/rust-GSL/commit/3b4c6e4856967e169af8fd2b74cb9052c93663f2) for example). 😆 Please clean it up then I'll make a last review.

Hmm, I think that particular commit was when I merged in some of your inline edits... Anyway, I just squashed all the commits. Should be good now.

GuillaumeGomez commented:
Thanks!
